Added TGA, BMP and PSD readers.  Added TGA and BMP writers.

ggarra13 opened the pull request:
They are very light as they rely on two auxiliary .h files (stb_image.h and stb_image_writer.h).

darbyjohnston commented on CMakeLists.txt:
```diff
@@ -42,6 +42,8 @@ set(TLRENDER_QT5 FALSE CACHE BOOL "Enable support for Qt5")
 set(TLRENDER_PROGRAMS TRUE CACHE BOOL "Build programs")
 set(TLRENDER_EXAMPLES TRUE CACHE BOOL "Build examples")
 set(TLRENDER_TESTS TRUE CACHE BOOL "Build tests")
+set(TLRENDER_GCOV FALSE CACHE BOOL "Enable gcov code coverage")
```
Could you check that you are synced with that latest? These coverage changes don't seem to be related to the PR.

ggarra13 replied:
Fixed.

darbyjohnston commented on lib/tlIO/STB.h:
```diff
@@ -0,0 +1,114 @@
+                uint16_t layer) override;
+        };
+
+        //! SGI writer.
```
SGI -> STB

darbyjohnston commented on lib/tlIO/STBRead.cpp:
```diff
@@ -0,0 +1,151 @@
+                    
+                    stbi_set_flip_vertically_on_load(1);
+
+                    int x, y, n;
```
It would be good to initialize these variables to zero.

darbyjohnston commented on lib/tlIO/STBRead.cpp:
```diff
@@ -0,0 +1,151 @@
+                File(const std::string& fileName, const file::MemoryRead* memory)
+                {
+                    int w = 0, h = 0, n = 0, bits = 8;
+                    int res = stbi_info(fileName.c_str(), &w, &h, &n);
```
Does STB have the ability to read from memory and files?

ggarra13 replied:
Yes.  I added new code to handle it now.

darbyjohnston reviewed (changes requested):
Looks like an interesting addition, though to be honest I was hoping to leave some of the older formats like TGA behind (DJV has TGA support which I did not port over to tlRender). On the other hand it's not a huge amount of code which is good.

My main concern would be if STB handles reading images from memory? That's necessary for the .otioz support which is essential. Also tests would need to be added similar to the other image formats in "tests\tlIOTest".

ggarra13 commented on lib/tlIO/CMakeLists.txt:
```diff
@@ -1,4 +1,6 @@
 set(HEADERS
+    stb_image.h
```
I moved the .h files to deps/stb.

ggarra13 commented on lib/tlIO/STBWrite.cpp:
```diff
@@ -0,0 +1,87 @@
+                        stbi_write_tga(fileName.c_str(), info.size.w,
+                                       info.size.h, comp, image->getData());
+                    else if (string::compareNoCase(ext, ".bmp"))
+                        stbi_write_bmp(fileName.c_str(), info.size.w,
```
No.  There's currently no library that handles writing PSD files, unfortunately.

ggarra13 commented:
> My main concern would be if STB handles reading images from memory? That's necessary for the .otioz support which is essential. Also tests would need to be added similar to the other image formats in "tests\tlIOTest".

It does handle images from memory.  I added code for it now, but I didn't know how to test it.  I also added a STBTest.cpp/.h based on the SGITest but also didn't know how to test it.  Can you test them and see if they are okay?

darbyjohnston commented:
You can run the tests like:
* make test (when using the CMake Makefile generator)
* run the test binary directly: tests/tltest/tltest

To test reading from memory you can use an .otioz file. However the "otioconvert" utility doesn't support image sequences yet, but I have a PR to fix that:
https://github.com/AcademySoftwareFoundation/OpenTimelineIO/pull/1627

darbyjohnston commented on lib/tlIO/STBWrite.cpp:
```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2021-2023 Darby Johnston
+// All rights reserved.
+
+#define STBIW_WINDOWS_UTF8
```
Just checking about these defines, they are different for reading and writing? The write doesn't need "STBI_NO_JPEG" and read doesn't need "STBIW_WINDOWS_UTF8"?

ggarra13 replied:
The NO_JPEG, etc. don't do anything in the stb_image_write.h file.  However you are right that I missed the UTF8 define for stb_image.h.  I fixed it (Good catch!).

darbyjohnston commented on lib/tlIO/STBRead.cpp:
```diff
@@ -0,0 +1,202 @@
+            {
+            public:
+                File(const std::string& fileName, const file::MemoryRead* memory)
+                    {
```
It looks like this block of code got extra indentation, the opening and closing brackets are not aligned.

ggarra13 replied:
Fixed.

darbyjohnston commented:
Thanks!
